controllers/pkg/reconcilers/bootstrap/reconciler.go

```diff
@@ -0,0 +1,277 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bootstrap
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	porchv1alpha1 "github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
+	porchconfigv1alpha1 "github.com/GoogleContainerTools/kpt/porch/api/porchconfig/v1alpha1"
+	"github.com/go-logr/logr"
+	"github.com/nephio-project/nephio/controllers/pkg/cluster"
+	ctrlconfig "github.com/nephio-project/nephio/controllers/pkg/reconcilers/config"
+	"github.com/nephio-project/nephio/controllers/pkg/resource"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	"k8s.io/apimachinery/pkg/util/yaml"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/kio/filters"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+)
+
+/*
+func init() {
+	controllers.Register("bootstrap", &reconciler{})
+}
+*/
+
+//+kubebuilder:rbac:groups="*",resources=secrets,verbs=get;list;watch
+//+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/status,verbs=get
+//+kubebuilder:rbac:groups=porch.kpt.dev,resources=packagerevisions,verbs=get;list;watch
+//+kubebuilder:rbac:groups=porch.kpt.dev,resources=packagerevisions/status,verbs=get
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *Reconciler) Setup(mgr ctrl.Manager, cfg *ctrlconfig.ControllerConfig) (map[schema.GroupVersionKind]chan event.GenericEvent, error) {
+	r.Client = mgr.GetClient()
+	r.porchClient = cfg.PorchClient
+
+	return nil, ctrl.NewControllerManagedBy(mgr).
+		Named("BootstrapController").
+		For(&corev1.Secret{}).
+		Complete(r)
+}
+
+type Reconciler struct {
+	client.Client
+	porchClient client.Client
+
+	l logr.Logger
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.l = log.FromContext(ctx)
+
+	cr := &corev1.Secret{}
+	if err := r.Get(ctx, req.NamespacedName, cr); err != nil {
+		// if the resource no longer exists the reconcile loop is done
+		if resource.IgnoreNotFound(err) != nil {
+			msg := "cannot get resource"
+			r.l.Error(err, msg)
+			return ctrl.Result{}, errors.Wrap(resource.IgnoreNotFound(err), msg)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// if the secret is being deleted dont do anything for now
+	if cr.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	// this branch handles installing the secrets to the remote cluster
+	if cr.GetNamespace() == "config-management-system" {
```

Comment:
This controller is running on the management cluster, isn't it?

So, the Secret is created in the "config-management-system" namespace of the management cluster (the local cluster)? Shouldn't it be in the same namespace as the Cluster resource? I thought the CS secret just had to live in the config-management-system namespace on the remote cluster.

Reply:
let me clarify:
- there is a token in the repo manifest that ask for a token for config-sync in ns config-system-management. This is applied to the management cluster but it is not used here. It will be picked up by the bootstrap controller to apply to the remote cluster.

The alternative would have been that I had to add the cluster specific handling in the token controller and I decided not to do this for not. So the code to deal with connecting to remote clusters is only in the bootstrap controller.

Reply:
ok. I think we can go for this with R1 but I foresee issues - for example, we should support multiple namespaces which could have the same names for repos, and we could get collisions here.